### PR TITLE
[Windows SEH] Fix crash on empty seh block

### LINKED ISSUE #105813: clang-cl.exe 19.1.0 and 20.0.0 hangs non-deterministically
My clang-cl.exe hangs for the following source code. In about 1 out of 10 runs the compilation is successful.

Environment: Windows 10 64-bit, Visual Studio 2022
Clang 19.1.0-rc2, prebuilt (LLVM-19.1.0-rc2-Windows-X64.tar.xz)
Clang 20.0.0, 64-bit, local debug build

`clang-cl.exe /nologo /MTd /EHa main.cxx`

------------------------------------------------------------------------------

[main.cxx]

```
#include <map>
#include <cassert>

void Foo (std::map<int,int> const &mapFoo,
          std::size_t const bar)
{
  auto cur = mapFoo.find(0);
  assert(bar>=0);
}

int main ()
{
  std::map<int,int> M;
  
  Foo (M, 7);
}
```

------------------------------------------------------------------------------

e:\test>clang-cl.exe /nologo /MTd /EHa main.cxx 
Assertion failed: !NodePtr->isKnownSentinel(), file E:\llvm\llvm-project\llvm\include\llvm/ADT/ilist_iterator.h, line 168
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: E:\\llvm\\20.0.0\\x86_64\\bin\\clang-cl.exe -cc1 -triple x86_64-pc-windows-msvc19.41.34120 -emit-obj -mincremental-linker-compatible -dumpdir a- -disable-free -clear-ast-before-backend -main-file-name main.cxx -mrelocation-model pic -pic-level 2 -mframe-pointer=none -relaxed-aliasing -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -fms-volatile -funwind-tables=2 -target-cpu x86-64 -mllvm -x86-asm-syntax=intel -tune-cpu generic -D_DEBUG -D_MT -flto-visibility-public-std --dependent-lib=libcmtd --dependent-lib=oldnames -stack-protector 2 -fcxx-exceptions -fexceptions -fasync-exceptions -fdiagnostics-format msvc -fdebug-compilation-dir=e:\\test -fcoverage-compilation-dir=e:\\test -resource-dir E:\\llvm\\20.0.0\\x86_64\\lib\\clang\\20 -internal-isystem E:\\llvm\\20.0.0\\x86_64\\lib\\clang\\20\\include -internal-isystem "C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\\VC\\Tools\\MSVC\\14.41.34120\\include" -internal-isystem "C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\\VC\\Tools\\MSVC\\14.41.34120\\atlmfc\\include" -internal-isystem "C:\\Program Files (x86)\\Windows Kits\\10\\Include\\10.0.22000.0\\ucrt" -internal-isystem "C:\\Program Files (x86)\\Windows Kits\\10\\Include\\10.0.22000.0\\shared" -internal-isystem "C:\\Program Files (x86)\\Windows Kits\\10\\Include\\10.0.22000.0\\um" -internal-isystem "C:\\Program Files (x86)\\Windows Kits\\10\\Include\\10.0.22000.0\\winrt" -internal-isystem "C:\\Program Files (x86)\\Windows Kits\\10\\Include\\10.0.22000.0\\cppwinrt" -fdeprecated-macro -ferror-limit 19 -fno-use-cxa-atexit -fms-extensions -fms-compatibility -fms-compatibility-version=19.41.34120 -std=c++14 -fskip-odr-check-in-gmf -fdelayed-template-parsing -faddrsig -o c:\\Temp\\main-7df83c.obj -x c++ main.cxx
1.	<eof> parser at end of file
2.	Code generation
3.	Running pass 'Function Pass Manager' on module 'main.cxx'.
4.	Running pass 'X86 DAG->DAG Instruction Selection' on function '@"?Foo@@YAXAEBV?$map@HHU?$less@H@std@@V?$allocator@U?$pair@$$CBHH@std@@@2@@std@@_K@Z"'
Exception Code: 0x80000003
 #0 0x00007ff6271432dc HandleAbort E:\llvm\llvm-project\llvm\lib\Support\Windows\Signals.inc:429:0
 #1 0x00007ffdbba98e05 (C:\WINDOWS\SYSTEM32\ucrtbased.dll+0xa8e05)
 #2 0x00007ffdbba9ab29 (C:\WINDOWS\SYSTEM32\ucrtbased.dll+0xaab29)
 #3 0x00007ffdbbaa094f (C:\WINDOWS\SYSTEM32\ucrtbased.dll+0xb094f)
 #4 0x00007ffdbba9e881 (C:\WINDOWS\SYSTEM32\ucrtbased.dll+0xae881)
 #5 0x00007ffdbbaa158f (C:\WINDOWS\SYSTEM32\ucrtbased.dll+0xb158f)
 #6 0x00007ff6212a4609 llvm::ilist_iterator<struct llvm::ilist_detail::node_options<class llvm::MachineInstr, 1, 1, void, 0, void>, 0, 0>::operator*(void) const E:\llvm\llvm-project\llvm\include\llvm\ADT\ilist_iterator.h:168:0
 #7 0x00007ff6212b1ec7 llvm::MachineInstrBundleIterator<class llvm::MachineInstr, 0>::operator*(void) const E:\llvm\llvm-project\llvm\include\llvm\CodeGen\MachineInstrBundleIterator.h:178:0
 #8 0x00007ff6298c137e llvm::SelectionDAGISel::reportIPToStateForBlocks(class llvm::MachineFunction *) E:\llvm\llvm-project\llvm\lib\CodeGen\SelectionDAG\SelectionDAGISel.cpp:1435:0
 #9 0x00007ff6298c3a04 llvm::SelectionDAGISel::SelectAllBasicBlocks(class llvm::Function const &) E:\llvm\llvm-project\llvm\lib\CodeGen\SelectionDAG\SelectionDAGISel.cpp:1857:0
#10 0x00007ff6298b4762 llvm::SelectionDAGISel::runOnMachineFunction(class llvm::MachineFunction &) E:\llvm\llvm-project\llvm\lib\CodeGen\SelectionDAG\SelectionDAGISel.cpp:614:0
#11 0x00007ff62466f30e `anonymous namespace'::X86DAGToDAGISel::runOnMachineFunction E:\llvm\llvm-project\llvm\lib\Target\X86\X86ISelDAGToDAG.cpp:191:0
#12 0x00007ff6298ca88d llvm::SelectionDAGISelLegacy::runOnMachineFunction(class llvm::MachineFunction &) E:\llvm\llvm-project\llvm\lib\CodeGen\SelectionDAG\SelectionDAGISel.cpp:374:0
#13 0x00007ff625954ab1 llvm::MachineFunctionPass::runOnFunction(class llvm::Function &) E:\llvm\llvm-project\llvm\lib\CodeGen\MachineFunctionPass.cpp:94:0
#14 0x00007ff62628c0f2 llvm::FPPassManager::runOnFunction(class llvm::Function &) E:\llvm\llvm-project\llvm\lib\IR\LegacyPassManager.cpp:1408:0
#15 0x00007ff62628c4e2 llvm::FPPassManager::runOnModule(class llvm::Module &) E:\llvm\llvm-project\llvm\lib\IR\LegacyPassManager.cpp:1454:0
#16 0x00007ff62628f358 `anonymous namespace'::MPPassManager::runOnModule E:\llvm\llvm-project\llvm\lib\IR\LegacyPassManager.cpp:1523:0
#17 0x00007ff626290279 llvm::legacy::PassManagerImpl::run(class llvm::Module &) E:\llvm\llvm-project\llvm\lib\IR\LegacyPassManager.cpp:541:0
#18 0x00007ff62628553c llvm::legacy::PassManager::run(class llvm::Module &) E:\llvm\llvm-project\llvm\lib\IR\LegacyPassManager.cpp:1651:0
#19 0x00007ff627ba3166 `anonymous namespace'::EmitAssemblyHelper::RunCodegenPipeline E:\llvm\llvm-project\clang\lib\CodeGen\BackendUtil.cpp:1152:0
#20 0x00007ff627ba3777 `anonymous namespace'::EmitAssemblyHelper::EmitAssembly E:\llvm\llvm-project\clang\lib\CodeGen\BackendUtil.cpp:1174:0
#21 0x00007ff627b9ef09 clang::EmitBackendOutput(class clang::DiagnosticsEngine &, class clang::HeaderSearchOptions const &, class clang::CodeGenOptions const &, class clang::TargetOptions const &, class clang::LangOptions const &, class llvm::StringRef, class llvm::Module *, enum clang::BackendAction, class llvm::IntrusiveRefCntPtr<class llvm::vfs::FileSystem>, class std::unique_ptr<class llvm::raw_pwrite_stream, struct std::default_delete<class llvm::raw_pwrite_stream>>, class clang::BackendConsumer *) E:\llvm\llvm-project\clang\lib\CodeGen\BackendUtil.cpp:1335:0
#22 0x00007ff6287cef12 clang::BackendConsumer::HandleTranslationUnit(class clang::ASTContext &) E:\llvm\llvm-project\clang\lib\CodeGen\CodeGenAction.cpp:354:0
#23 0x00007ff62d31de05 clang::ParseAST(class clang::Sema &, bool, bool) E:\llvm\llvm-project\clang\lib\Parse\ParseAST.cpp:191:0
#24 0x00007ff628cc20d2 clang::ASTFrontendAction::ExecuteAction(void) E:\llvm\llvm-project\clang\lib\Frontend\FrontendAction.cpp:1192:0
#25 0x00007ff6287cc007 clang::CodeGenAction::ExecuteAction(void) E:\llvm\llvm-project\clang\lib\CodeGen\CodeGenAction.cpp:1144:0
#26 0x00007ff628cc19bf clang::FrontendAction::Execute(void) E:\llvm\llvm-project\clang\lib\Frontend\FrontendAction.cpp:1078:0
#27 0x00007ff628c80e36 clang::CompilerInstance::ExecuteAction(class clang::FrontendAction &) E:\llvm\llvm-project\clang\lib\Frontend\CompilerInstance.cpp:1061:0
#28 0x00007ff628f6ab85 clang::ExecuteCompilerInvocation(class clang::CompilerInstance *) E:\llvm\llvm-project\clang\lib\FrontendTool\ExecuteCompilerInvocation.cpp:280:0
#29 0x00007ff62120205d cc1_main(class llvm::ArrayRef<char const *>, char const *, void *) E:\llvm\llvm-project\clang\tools\driver\cc1_main.cpp:285:0
#30 0x00007ff6211ebcfe ExecuteCC1Tool E:\llvm\llvm-project\clang\tools\driver\driver.cpp:215:0
#31 0x00007ff6211ec513 clang_main(int, char **, struct llvm::ToolContext const &) E:\llvm\llvm-project\clang\tools\driver\driver.cpp:256:0
#32 0x00007ff621238096 main E:\llvm\build_vs17_x64\tools\clang\tools\driver\clang-driver.cpp:17:0
#33 0x00007ff631ee1279 invoke_main D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:79:0
#34 0x00007ff631ee1162 __scrt_common_main_seh D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288:0
#35 0x00007ff631ee101e __scrt_common_main D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:331:0
#36 0x00007ff631ee130e mainCRTStartup D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_main.cpp:17:0
#37 0x00007ffddf3c7374 (C:\WINDOWS\System32\KERNEL32.DLL+0x17374)
#38 0x00007ffde011cc91 (C:\WINDOWS\SYSTEM32\ntdll.dll+0x4cc91)
clang-cl: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 20.0.0git (https://github.com/llvm/llvm-project.git 27e5f505e5eee3da27c1515d6ed95d66fbe543ea)
Target: x86_64-pc-windows-msvc
Thread model: posix
InstalledDir: E:\llvm\20.0.0\x86_64\bin
Build config: +unoptimized, +assertions
clang-cl: note: diagnostic msg: 
********************

### PR DESCRIPTION
**Pull request by R-Goc**
Fixes https://github.com/llvm/llvm-project/issues/105813 and https://github.com/llvm/llvm-project/issues/106915.
Adds a check for the end of the iterator, which can be a sentinel. 
The issue was introduced in https://github.com/llvm/llvm-project/commit/0efe111365ae176671e01252d24028047d807a84 from what I can see, so along with the introduction of /EHa support.

### CODE REVIEW

**Comment by R-Goc** [comment]
I'm aware that clang-format may have to be run on this, not sure why my editor changed anything.

**Comment by R-Goc** [comment]
Thanks to @glebov-andrey, @asl and @jrtc27.

**Comment by jrtc27** [comment]
It looks like you ran clang-format on the file rather than the diff?

**Comment by jrtc27** [comment]
This also would ideally have a test case.

**Comment by jrtc27** [comment]
Not really. Ideally there would be a minimal reproducer in llvm/test.

**Comment by R-Goc** [comment]
I'm having trouble getting the reproducer to be an .ll or .bc file. It seems even less deterministic in that form. For example debug llc doesn't hang on it.  Bugpoint fails to emit a proper file as well. Currently there are 3 known reproductions, 2 of which are very delicate and fail with seemingly unrelated options. The one that basically always fails is a 350k loc part of protobufs.

**Comment by jrtc27** [comment]
Reduced: https://godbolt.org/z/7hhsxeqj5

**Comment by R-Goc** [comment]
Out of curiosity how did you do it? Manually?

**Comment by jrtc27** [comment]
I ran the C++ through creduce to cut it down to something llvm-reduce wouldn't crash on (it seemed to quickly bail on an invoke IIRC, perhaps not adapted to Windows-style exceptions), and then ran it through llvm-reduce to get a minimal IR test.

**Comment by R-Goc** [comment]
Okay. In terms of naming I'm not sure. Does wineh-unclosed-scope.ll work?

**Comment by R-Goc** [comment]
What is left to do to get this merged?

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp [hunk 1]
```diff
@@ -4429,4 +4434,4 @@ void SelectionDAGISel::CannotYetSelect(SDNode *N) {
       Msg << "unknown intrinsic #" << iid;
   }
   report_fatal_error(Twine(msg));
-}
+}
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:4437]: Missing newline at end of file

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp [hunk 2]
```diff
@@ -1476,6 +1476,11 @@ void SelectionDAGISel::reportIPToStateForBlocks(MachineFunction *MF) {
     if (BB->getFirstMayFaultInst()) {
       // Report IP range only for blocks with Faulty inst
       auto MBBb = MBB.getFirstNonPHI();
+
+      // Avoids attempting to dereference a sentintel which fails an assert
+      if (MBBb == MBB.instr_end())
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:1481]: Just use .end, not instr_end

File: llvm/test/CodeGen/WinEH/wineh-empty-seh-scope.ll [hunk 3]
```diff
@@ -0,0 +1,14 @@
+; RUN: llc -mtriple=x86_64-pc-windows-msvc19.41.34120 %s
```
Comment (arsenm) [inline llvm/test/CodeGen/WinEH/wineh-empty-seh-scope.ll:1]: Add FileCheck and generated checks with update_llc_test_checks

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp [hunk 4]
```diff
@@ -1476,6 +1476,11 @@ void SelectionDAGISel::reportIPToStateForBlocks(MachineFunction *MF) {
     if (BB->getFirstMayFaultInst()) {
       // Report IP range only for blocks with Faulty inst
       auto MBBb = MBB.getFirstNonPHI();
+
+      // Avoids attempting to dereference a sentintel which fails an assert
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:1480]: Don't need this comment

File: llvm/test/CodeGen/WinEH/wineh-empty-seh-scope.ll (hunk 3 above)
Comment (arsenm) [reply llvm/test/CodeGen/WinEH/wineh-empty-seh-scope.ll:1]: Still prefer tests with checks to tests that just compile

Comment (R-Goc) [reply llvm/test/CodeGen/WinEH/wineh-empty-seh-scope.ll:1]: Didn't touch this yet. Never used FileCheck nor that script so I'll need to read the docs.

Comment (arsenm) [reply llvm/test/CodeGen/WinEH/wineh-empty-seh-scope.ll:1]: Just add | FileCheck %s and then run update_llc_test_checks on the file

**Review by arsenm — APPROVED** [review]

**Comment by R-Goc** [comment]
Should I rebase and squash now?

**Comment by R-Goc** [comment]
A lot of commits seem to be waiting for linux workers. Is this normal?

**Comment by arsenm** [comment]
> A lot of commits seem to be waiting for linux workers. Is this normal?

Yes

**Comment by arsenm** [comment]
Can you reword the commit title to be more specific about empty blocks

**Comment by R-Goc** [comment]
Would this be a good candidate to backport to LLVM 19?

**Comment by arsenm** [comment]
> Would this be a good candidate to backport to LLVM 19?

Yes

**Comment by mstorsjo** [comment]
@R-Goc See https://github.com/llvm/llvm-project/blob/main/.github/workflows/email-check.yaml#L29-L39, https://discourse.llvm.org/t/hidden-emails-on-github-should-we-do-something-about-it and https://github.com/settings/emails. (There's an automation workflow that's supposed to detect this and remind you about it, but it seems to not have run here.)

**Pull request merged**